File: golang/datastructure/skiplist/skiplist.go
// 1206. 设计跳表 https://leetcode.cn/problems/design-skiplist/
package main

import (
	"fmt"
	"math/rand"
)

const (
	p        = 0.25
	maxLevel = 32
)

type Node struct {
	val     int
	forward []*Node
}

type Skiplist struct {
	head  *Node
	level int
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func randLevel() int {
	lv := 1
	for lv < maxLevel && rand.Float64() < p {
		lv++
	}
	return lv
}

func Constructor() Skiplist {
	return Skiplist{&Node{-1, make([]*Node, maxLevel)}, 0}
}

func (s *Skiplist) Search(target int) bool {
	cur := s.head
	for i := s.level - 1; i >= 0; i-- {
		// 找到第i层比target小且最接近的节点
		for cur.forward[i] != nil && cur.forward[i].val < target {
			cur = cur.forward[i]
		}
	}
	cur = cur.forward[0]
	return cur != nil && cur.val == target
}

func (s *Skiplist) Add(num int) {
	lv := randLevel()
	cur, update := s.head, make([]*Node, lv)
	for i := 0; i < lv; i++ {
		update[i] = s.head
	}
	for i := s.level - 1; i >= 0; i-- {
		for cur.forward[i] != nil && cur.forward[i].val < num {
			cur = cur.forward[i]
		}
		if i < lv {
			update[i] = cur
		}
	}
	s.level = max(s.level, lv)
	newNode := &Node{num, make([]*Node, lv)}
	for i, node := range update {
		newNode.forward[i] = node.forward[i]
		node.forward[i] = newNode
	}
}

func (s *Skiplist) Erase(num int) bool {
	cur, update := s.head, make([]*Node, maxLevel)
	for i := s.level - 1; i >= 0; i-- {
		for cur.forward[i] != nil && cur.forward[i].val < num {
			cur = cur.forward[i]
		}
		update[i] = cur
	}
	cur = cur.forward[0]
	if cur == nil || cur.val != num {
		return false
	}
	for i := 0; i < s.level && update[i].forward[i] == cur; i++ {
		update[i].forward[i] = cur.forward[i]
	}
	for s.level > 0 && s.head.forward[s.level-1] == nil {
		s.level--
	}
	return true
}

func main() {
	s := Constructor()
	arr := []int{}
	for i := 0; i < 10; i++ {
		v := int(rand.Int31())
		arr = append(arr, v)
		s.Add(v)
	}
	for _, v := range arr {
		fmt.Printf("Add %d\n", v)
	}
	fmt.Printf("search %d: %v\n", arr[0], s.Search(arr[0]))
	fmt.Printf("erase %d: %v\n", arr[0], s.Erase(arr[0]))
	fmt.Printf("search %d: %v\n", arr[0], s.Search(arr[0]))
}

File: c/datastructure/list/skiplist/skiplist.c
//
// Created by john on 2022/10/7.
//

#include "skiplist.h"

static const int P = RAND_MAX >> 2;
static const int MAX_LEVEL = 32;

Node *nodeCreate(int val, int level) {
    Node *t = (Node *) malloc(sizeof(Node));
    t->val = val;
    t->size = level;
    t->forward = (Node **) malloc(level * sizeof(Node *));
    for (int i = 0; i < level; i++) {
        t->forward[i] = NULL;
    }
    return t;
}

Skiplist *skiplistCreate() {
    Skiplist *skiplist = (Skiplist *) malloc(sizeof(Skiplist));
    skiplist->level = 0;
    skiplist->head = nodeCreate(-1, MAX_LEVEL);
    srand(time(NULL));
    return skiplist;
}

static inline int randLevel() {
    int level = 1;
    while (level < MAX_LEVEL && rand() < P) {
        level++;
    }
    return level;
}

bool skiplistSearch(Skiplist *obj, int target) {
    Node *cur = obj->head;
    for (int i = obj->level - 1; i >= 0; i--) {
        while (cur->forward[i] && cur->forward[i]->val < target) {
            cur = cur->forward[i];
        }
    }
    cur = cur->forward[0];
    return cur && cur->val == target;
}

void skiplistAdd(Skiplist *obj, int num) {
    int lv = randLevel();
    Node *t = nodeCreate(num, lv);
    Node *cur = obj->head;
    Node *pre[MAX_LEVEL];
    for (int i = MAX_LEVEL - 1; i >= 0; i--) {
        while (cur->forward[i] && cur->forward[i]->val < num) {
            cur = cur->forward[i];
        }
        if (i < lv) {
            pre[i] = cur;
        }
    }
    for (int i = lv - 1; i >= 0; i--) {
        t->forward[i] = pre[i]->forward[i];
        pre[i]->forward[i] = t;
    }
    obj->level = max(obj->level, lv);
}

void nodeFree(Node *node) {
    if (node->forward) {
        free(node->forward);
        node->forward = NULL;
        node->size = 0;
        node->val = 0;
    }
    free(node);
}

bool skiplistErase(Skiplist *obj, int num) {
    Node *cur = obj->head;
    Node *pre[MAX_LEVEL];
    for (int i = MAX_LEVEL - 1; i >= 0; i--) {
        while (cur->forward[i] && cur->forward[i]->val < num) {
            cur = cur->forward[i];
        }
        pre[i] = cur;
    }
    cur = cur->forward[0];
    if (!cur || cur->val != num) {
        return false;
    }
    for (int i = cur->size - 1; i >= 0; i--) {
        pre[i]->forward[i] = cur->forward[i];
    }
    if (obj->level > 0 && !obj->head->forward[obj->level - 1]) {
        obj->level--;
    }
    nodeFree(cur);
    return true;
}


void skiplistFree(Skiplist *obj) {
    for (Node *cur = obj->head; cur;) {
        Node *t = cur;
        cur = cur->forward[0];
        nodeFree(t);
    }
    free(obj);
}